spec/main.tex

```diff
@@ -0,0 +1,1398 @@
+Computing $s^x \bmod N_1$ and $s^x \bmod N_2$ will, in general, be faster than computing $s^x \bmod N$ directly because (1)~$N_1, N_2$ are shorter than~$N$, and (2)~assuming the factorizations of $N_1, N_2$ are known, we can reduce the exponent~$x$ modulo $\phi(N_1)$ (resp.,~$\phi(N_2)$) before performing the respective exponentiations.
+Namely, we can use the fact that, e.g., 
+\[s^x \bmod N_1 = s^{x \bmod \phi(N_1)} \bmod N_1\]
+(assuming $\gcd(s, N_1)=1$.)
+
+\medskip\noindent{\bf Combining the results.}
+Let $\beta=N_1^{-1} \bmod N_2$. %(Note this can be computed as part of a preprocessing step.) 
+If we have computed $r_1=s^x \bmod N_1$ and $r_2=s^x \bmod N_2$, we can compute $s^x \bmod N$ as
+\[r_1+\left((r_2-r_1) \cdot \beta \bmod N_2\right) \cdot N_1 .\]
+(Note that no reduction modulo $N$ is needed, and the result will already be in the correct range.)
+To see that this gives the correct answer, note that 
+\begin{eqnarray*}
+& r_1+\left((r_2-r_1) \cdot \beta \bmod N_2\right) \cdot N_1 =  r_1 \bmod N_1 &\\
+& r_1+\left((r_2-r_1) \cdot \beta \bmod N_2\right) \cdot N_1 = r_1+\left((r_2-r_1) \cdot N_1^{-1} \right) \cdot N_1 = r_2 \bmod N_2. &
+\end{eqnarray*}
+%then apply the Chinese remainder theorem.
+
+
+\medskip
+In our context, the case of interest is when the modulus is $N^2=p^2 q^2$ and the factors $p, q$ are known.
+Algorithm~\ref{alg:CRT} shows a complete algorithm for exponentiation modulo~$N^2$ in that case.
+
+\begin{algorithm}
+\caption{Computing $s^x \bmod N^2$, where $N^2=p^2q^2$ with $p, q$ distinct primes and $\gcd(s,N^2)=1$}\label{alg:CRT}
+\begin{algorithmic}[1]
+\State $\beta:=p^{-2} \bmod q^2$ \;\;\; //  this can be computed in a preprocessing step
+\State $\phi_1 = p \cdot (p-1), \; \phi_2 := q \cdot (q-1)$
+\State $x_1:= x \bmod \phi_1, \; x_2:=x \bmod \phi_2$
+\State $s_1 := s \bmod p^2, \; s_2 := s \bmod q^2$
+\State $r_1:= s_1^{x_1} \bmod p^2, \; 
+r_2 := s_2^{x_2} \bmod q^2$
+\State $\res := r_1+((r_2-r_1)\cdot \beta \bmod q^2) \cdot p^2$
+\State \Return \res
+\end{algorithmic}
+\end{algorithm}
+
+
+\subsection{Paillier Encryption Scheme}
+We include a description of the algorithms that constitute the Paillier encryption scheme. 
+%Note, however, that we call an external library for these algorithms, and do not implement them ourselves.
+\begin{enumerate}
+    \item $\textsf{keygen}$ generates a private key $\sk$ consisting of two safe primes, with the public key being their product~$N$.
+    
+    
+    \item 
+		$\enc_N(M; r)$ 
+		encrypts a message $M \in \{-(N-1)/2, \ldots,  (N-1)/2\}$ 
+		using randomness $r \in \Z_N^*$ 
+		and Paillier public key~$N$. 
+		This produces a ciphertext $C \in \Z^*_{N^2}$.
+		Encryption checks that $\gcd(r, N)=1$ (and raises an error if not), and then computes
+    
+		$$\enc_N(M; r) := (1 + M \cdot N) \cdot r^N \bmod N^2$$ 
+
+        We also provide a function $\enc_N(M)$ that samples uniform $r \in \Z_N^*$ and returns~$\enc_N(M; r)$.
```

Comment:
Mixing tabs and spaces

Reply:
PR did not introduce that, it existed before

Reply:
Oh actually there were no tabs before. But my editor thought the tabbing would be appropriate choice for some reason. Maybe it found tabs in some other place

Reply:
Fixed